### ISSUE TITLE
Custom Provider in Local Paths does not support version

### ISSUE THREAD

**Issue by Techcadia** [hashicorp/terraform#16874]
So I have been trying to make this work and I finally have decided its time to open a ticket. Here is my use case. I develop a provider for teamcity. I want to keep a specific version pinned so I don't break our remote state with new version additions and deletions while working on the provider. Currently I have to remember to go and checkout and compile the version. From reading all the docs I should be able to place the bin files with version number in my home directory and then set the version. This does not work as intended. At least for providers that do not have a folder in the releases s3 bucket since it throws an error.

Thanks for the assistance.

### Terraform Version

```
v0.11.0
```

### Terraform Configuration Files

```hcl
...
provider "teamcity" {
  url = "http://localhost"
  version = ">= 0.1.3"
}
```

### Debug Output

https://gist.github.com/Techcadia/2dfc88be06af3608f11aead032a9313e (Default Init)
https://gist.github.com/Techcadia/8914456a19610ef3092fe312cbeb5877 (Get Plugins Disabled)

### Expected Behavior

The Provider should not be found at https://releases.hashicorp.com/terraform-provider-teamcity/ but the local file should be used as it should override the public release that does not exist.

### Actual Behavior

Failure because it does not exist in the S3 releases bucket. Which is true because this is a third party provider.

or if get-plugins=false

```
The following provider constraints are not met by the currently-installed
provider plugins:

* teamcity (>= 0.1.3)
```

### Steps to Reproduce

Compile https://github.com/Cardfree/terraform-provider-teamcity/releases
Move Compiled version
```
mv $GOPATH/bin/terraform-provider-teamicty to ~/.terraform.d/plugins/darwin_amd64/terraform-provider-teamicty_v0.1.3
terraform init
```
Failure occurs.

**Comment by apparentlymart** [comment]
Hi @Techcadia! Sorry this isn't working as expected.

Unfortunately I wasn't able to reproduce this on my system :confounded: so it seems like there's something tricky/subtle going on.

(Note that I'm on Linux, so `linux_amd64` in my log is equivalent to `darwin_amd64` in yours)

```
...
2017/12/07 10:56:29 [DEBUG] checking for provider in "."
2017/12/07 10:56:29 [DEBUG] checking for provider in "/home/apparentlymart/devel/terraform/bin"
2017/12/07 10:56:29 [DEBUG] checking for provider in "/home/apparentlymart/.terraform.d/plugins"
2017/12/07 10:56:29 [DEBUG] checking for provider in "/home/apparentlymart/.terraform.d/plugins/linux_amd64"
2017/12/07 10:56:29 [DEBUG] found provider "terraform-provider-teamcity_v0.1.3"
2017/12/07 10:56:29 [DEBUG] found valid plugin: "teamcity"
2017/12/07 10:56:29 [DEBUG] checking for provisioner in "."
...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

Your log output suggests that Terraform _is_ finding the plugin binary, but for some reason it is not accepting it as a match for the constraint `>= 0.1.3` even though you have the version constraint in the name. I _can_ reproduce the behavior you saw if I change the constraint to something that doesn't match, like `version = ">= 0.2.3"`, which I think confirms the fact that this is a constraint mismatch for you, but I don't yet have a theory as to why that would be.

Does the behavior change if you alter the version constraint to some other, more "liberal" strings, like these?:

* `>= 0.2.0`
* `>= 0.0.0`

I'm curious to try to determine whether it's just _never_ matching or if Terraform is just misinterpreting the specific constraint you gave here.

**Comment by Techcadia** [comment]
@apparentlymart Thanks for the comment I don't see a note about if you ran with or without the `-get-plugins=false` so I tested both ways.
```
2017/12/07 12:36:37 [DEBUG] checking for provider in "/Users/techcadia/.terraform.d/plugins/darwin_amd64"
2017/12/07 12:36:37 [DEBUG] found provider "terraform-provider-teamcity_v0.1.3"
2017/12/07 12:36:37 [DEBUG] found valid plugin: "teamcity"
2017/12/07 12:36:37 [DEBUG] plugin requirements: "teamcity"=">= 0.2.0"
... 403 Error again if '-get-plugins=true' if '-get-plugins=false' no error but same failure
2017/12/07 12:36:37 [ERROR] 1 error(s) occurred:
* no provider exists with the given name
```
Here is the `>= 0.0.0` run which did work 
https://gist.github.com/Techcadia/a8acf09219e79af38740e120144aad71

Here is direct version '0.1.3' which did not work either
https://gist.github.com/Techcadia/4016335bdbabe678f0c7f138014e5e05

**Comment by Techcadia** [comment]
So I just had a coworker that runs on linux validate this for me and he is seeing the same results as you. This works as intended in linux.

Looking through his code I see no error from S3 for the release request. So either the code confirms functionality before it makes the request here https://gist.github.com/Techcadia/2dfc88be06af3608f11aead032a9313e#file-gistfile1-txt-L34

**Comment by Techcadia** [comment]
So one interesting tidbit.

https://github.com/hashicorp/terraform/blob/v0.11.1/command/init.go#L313-L314
spew of missing and internal
```
(discovery.PluginRequirements) (len=1) {
 (string) (len=8) "teamcity": (*discovery.PluginConstraints)(0xc4204324e0)({
  Versions: (discovery.Constraints) 0.1.3,
  SHA256: ([]uint8) <nil>
 })
}
(map[string]terraform.ResourceProviderFactory) (len=1) {
 (string) (len=9) "terraform": (terraform.ResourceProviderFactory) 0x279a500
}
```

So it thinks the package does not exist on the local file system and proceeds to go check external sources.

**Comment by jbardin** [comment]
Hi @Techcadia,

I verified that this is working as expected on darwin under normal circumstances. I'm still trying to come up with a plausible scenario that leads to the log output you've provided. While it's a long shot, the only way I've gotten the logs you've shown is that the text `"teamcity"` in the config doesn't exactly match the `teamcity` in the filename via some unicode homoglyphs. You could try carefully retyping the provider in the config and rename the binary by hand and see if that's the case.

**Comment by Techcadia** [comment]
@jbardin @apparentlymart Thanks guys for the help. I took some time to Trace this down and I have a PR with a suggested fix to keep others from banging their heads at the wall like I did.

Super simple mistake I made. I failed to check my .terraformrc file to see if I had overridden the Provider (plugin) Which I had. The part that had me chase the wrong rabbit down the hole was the version.

Here is my .terraformrc
```
providers {
  teamcity = "${GOPATH}/bin/terraform-provider-teamcity"
}
```

https://github.com/hashicorp/terraform/pull/16882

Let me know if you guys have any questions I am going to mark this as resolved since there never was a bug just me missing a key piece.
